Second PR - HON game (infinite timers)

abhishek-tripathi-yral opened the pull request:
Merge this only after #352 

- timers for new yral game - timer starts from the time when the first bet starts. 
- slot_id  - uses NewSlotId - which is u64 instead of u8

abhishek-tripathi-yral committed:
feat: migration PR only

abhishek-tripathi-yral committed:
fix: make dfx build happen

abhishek-tripathi-yral committed:
chore: ignore on main

abhishek-tripathi-yral committed:
feat: second PR

abhishek-tripathi-yral committed:
feat: pass one more test

abhishek-tripathi-yral committed:
fix: update did files

abeeshake-yral commented:
Rebased on first PR . Please merge this only after merging first one.

abeeshake-yral commented:
on branch `abhi-yral-005`
 `hotornot_game_simulation_test_after_yral_game_3` has been put as `#[ignore]` for the branch. Feel free to run that locally. 
This is a counterpart test to `hotornot_game_simulation_test_2`

abhishek-tripathi-yral committed:
feat: hotornot_game_simulation_test_after_yral_game_3

abeeshake-yral commented:
more debugging related exploration here: https://gist.github.com/abeeshake-yral/ae98a6bf2028d7d2c6687a8684255b06

abhishek-tripathi-yral committed:
feat: slightly optimize timer logic

abeeshake-yral commented:
Betting starts when the first user bets on a given post. Betting slot is alloted thena and a timer for 1 hr is started.


## Here are two apporaches to implement the timer:

1. have a timer for each post which is being betted on
2. have a global timer and a queue of posts that are being betted on.
   1. iterate over the queue and check if the timer has expired.
   2. if the timer has expired, then tabulate the outcome for the post and remove the post from the queue.

We chose the second approach because it is more efficient.

### Here are the scenarios that can happen for the timers:

post_2_Z => means post 2 for the user with id Z

1. User A bets on a post_1_Z at 04:00  and the global timer for 1 hr is started (to be finished at 05:00)
2. User B bets on a post_1_Z at 04:01 and no timer is started. (since there should be ONLY one timer in the canister at a time)
3. User C bets on a post_2_Z at 04:02 and no timer is started. (since there should be ONLY one timer in the canister at a time)


When timer expires, post_1_Z is `tabulate_hot_or_not_outcome_for_post_slot_v1` and post_1_Z is removed from the queue. Next timer is queued for the next post (post_2_Z) in queue. 

Note that this second timer actually needs to run only for 1 minute since this timer is started at 05:00 and the expiry of this timer needs to be at 05:01.  Computationally, the function `tabulate_hot_or_not_outcome_for_post_slot_v1` after 1 hr for post_2_Z. 


### Here are the technical details for implementing the timer:

1. queue is best implemented in stable memory as `ic_stable_structures::btree_map::BTreeMap` because it has `pop_first` and `insert` methods. The Vec in ic_stable_structures doesn't have `pop_first`.
2. we also need to keep track of the time when the first bet was placed for each post. This is because timer expires 1 hr after the first bet on the given post is placed.
```rust
    pub bet_timer_posts: ic_stable_structures::btreemap::BTreeMap<(SystemTimeInMs, PostId), (), Memory>,
```

#### Here are the scenarios that can happen for the timers:

Think of these sequentially.

A1. when first bet on post_1_Z is placed, and bet_timer_posts is empty. We insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId) and `start_timer`.

A2. when second bet on post_1_Z is placed, we *check* if PostId(post_1_Z) exists in bet_timer_posts. if it doesn't, then we insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId). we don't ned to call `start_timer` here. Since there must be a global timer running. It was set in A1. 

A3. when first bet on post_2_Z is placed, we *check* if PostId(post_2_Z) exists in bet_timer_posts. It doesn't. So, we insert into bet_timer_posts a tuple of (SystemTimeInMs, PostId). we don't ned to call `start_timer` here.

A4. when timer for post_1_Z expires, we `tabulate_hot_or_not_outcome_for_post_slot_v1` for post_1_Z and remove it from bet_timer_posts. also, start the next timer (via `start_timer`) for the next post (post_2_Z) in the queue bet_timer_posts.

Now, in A2 above, to check if the posts has already been betted upon, in O(1) time, we need a separate BTreeMap<PostId, ()>. we call this map `first_bet_placed_at_hashmap`.
now this map needs to be updated always along with bet_timer_posts. at the cost of memory, we are optimising for lookup speed.

debjit-bhowal-rs committed:
feat: device id management functionality on canister (#360)

* push notif functionality added

* candid file updated

* bug fixes

* feat: finished device ID management functionality

* added back access control

* type fix

abhishek-tripathi-yral committed:
feat: yral feed which is merged version of hot or not and home feed (#356)

* feat: yral feed which is merged version of hot or not and home feed

* fix: yral feed no duplicates

assumption: the hot or not feed and home feed have non-intersectional
post_ids

* fix: one step migration test

abhishek-tripathi-yral committed:
Candid files for the HON game (48h)  (#371)

fix: candid files for hon game

komal-sai-yral committed:
issue fix possibly (#358)

abhishek-tripathi-yral committed:
wip: update did files

abhishek-tripathi-yral committed:
fix: do the migration in one PR only.

- serde gives the default values to skip fields. so there is no need for
doing two step migration

The Two step migration plan (earlier) was:

1. skip_deserializing in first PR
2. migrate data in second PR

Now, 1. step is taken care of by `Default::default()` since serde does
that for you.

abhishek-tripathi-yral committed:
fix: all_hot_or_not_bets_placed_v1

abhishek-tripathi-yral committed:
Revert "fix: all_hot_or_not_bets_placed_v1"

This reverts commit 2f1122d3e62bacaae7761c05de145755a64121f7.

abhishek-tripathi-yral committed:
fix: serde default for single step migration

abhishek-tripathi-yral committed:
Merged main into HON-infinite-timers

abhishek-tripathi-yral committed:
fix; updated did file

abhishek-tripathi-yral committed:
Merge branch 'main' into abhi-yral-005

abhishek-tripathi-yral committed:
fix: format and compile

abhishek-tripathi-yral committed:
fix: updated the concurrent bet logic from old fix

abhishek-tripathi-yral committed:
fix: remove the default timers

- these timers were created when the post was created. after moving to
  hot or not infinite game -> we don't need it

abhishek-tripathi-yral committed:
chore: ignore the old timer test

- old timer test fails because the logic is changed.
- new timer test hotornot_game_simulation_test_yral passes with new
  timer logic

abhishek-tripathi-yral committed:
Merge branch 'main' into abhi-yral-005

ravibazz commented on src/canister/individual_user_template/src/api/hot_or_not_bet/receive_bet_from_bet_makers_canister.rs:
```diff
@@ -80,6 +90,196 @@ fn receive_bet_from_bet_makers_canister_impl(
+            ..
+        } => {
+            // insert only the first bet in first_bet_placed_at_hashmap
+            if !canister_data
+                .first_bet_placed_at_hashmap
+                .contains_key(&place_bet_arg.post_id)
```
We should also compare the ongoing slot for in this check as if we the ongoing slot is different and we miss inserting that into the queue it would never be calculated.

ravibazz commented on src/canister/individual_user_template/src/data_model/mod.rs:
```diff
@@ -42,26 +47,59 @@ use kv_storage::AppStorage;
     pub configuration: IndividualUserConfiguration,
     pub follow_data: FollowData,
     pub known_principal_ids: KnownPrincipalMap,
+    // #[serde(skip)]
```
Should we remove this?

ravibazz commented on src/canister/individual_user_template/src/data_model/mod.rs:
```diff
@@ -82,34 +120,50 @@ pub struct CanisterData {
     #[serde(skip, default = "_default_success_history_v1")]
     pub success_history: ic_stable_structures::btreemap::BTreeMap<SuccessHistoryItemV1, (), Memory>,
+    // u64 is post_id, SystemTime refers to time when first_bet is placed
+    #[serde(skip, default = "_default_bet_timer_posts_queue")]
+    pub bet_timer_posts:
+        ic_stable_structures::btreemap::BTreeMap<(SystemTimeInMs, PostId), (), Memory>,
+    // this keeps track of when the first bet was placed.
+    #[serde(skip, default = "_default_bet_timer_first_bet_placed_at_map")]
+    pub first_bet_placed_at_hashmap:
+        ic_stable_structures::btreemap::BTreeMap<PostId, (SystemTimeInMs, NewSlotType), Memory>,
```
We can combine both of this into one. There is no need to have two maps if you think logically. Also we can store `PlaceBetArg, ()` directly.

ravibazz commented:
Closing this as this is no longer required.
